File: lib/data/local/tables/sync_meta.dart
import 'package:drift/drift.dart';

import '../converters.dart';

/// The columns every synchronised row carries.
///
/// Mixed into each domain table so the sync contract is defined once. drift has
/// no embedded-object concept, which is an improvement here: these land as flat
/// columns with no prefix, so a DAO can say `where deleted equals false`
/// directly.
///
/// Three rules travel with this mixin and are enforced by convention plus tests
/// rather than by the type system, so they are worth restating:
///
/// 1. **[version], [serverSeq] and [serverUpdatedAt] are server-owned.** A local
///    edit carries them forward untouched. Resetting [version] makes the next
///    upload conflict with itself, because the server compares what you send
///    against what it stored.
/// 2. **There is no `dirty` flag.** Dirtiness *is* `syncStatus != synced`. A
///    second boolean could disagree with the status and there would be no rule
///    for which one wins.
/// 3. **[createdAt] and [updatedAt] are device clocks, for display and local
///    ordering only.** They are never used to resolve a conflict -- see
///    constraint #3 in CLAUDE.md, and [serverUpdatedAt] for the authoritative
///    time.
mixin SyncMetaColumns on Table {
  /// Optimistic-concurrency version, assigned by the server.
  ///
  /// Sent back as `baseVersion` so the server can reject a write built on a
  /// stale read. Zero means the row has never been accepted by the server.
  IntColumn get version => integer().withDefault(const Constant(0))();

  /// The server's global change sequence for this row, and the download cursor.
  ///
  /// Null until the server has seen the row. Also the tombstone-purge key: a
  /// deleted row may only be removed once its [serverSeq] is below the
  /// slowest device's cursor.
  IntColumn get serverSeq => integer().nullable()();

  /// When the server last wrote this row, ISO-8601. Authoritative, unlike
  /// [updatedAt].
  TextColumn get serverUpdatedAt => text().nullable()();

  /// Whether this row matches the server.
  ///
  /// Also answers §98's `metadataSyncStatus`; see constraint #7.
  TextColumn get syncStatus => text().map(syncStatusConverter)();

  /// Soft-delete flag (§36). The row stays so the removal propagates.
  BoolColumn get deleted => boolean().withDefault(const Constant(false))();

  /// When the row was soft-deleted, epoch milliseconds.
  IntColumn get deletedAt => integer().nullable()();

  /// Who soft-deleted the row.
  TextColumn get deletedBy => text().nullable()();

  /// When the row was created on the originating device, epoch milliseconds.
  IntColumn get createdAt => integer()();

  /// When the row was last edited locally, epoch milliseconds.
  ///
  /// Load-bearing for one specific rule: the uploader captures this at send
  /// time and only marks the row synced if it still matches, so an edit made
  /// while the request was in flight is not silently discarded.
  IntColumn get updatedAt => integer()();

  /// Author's identity, stamped by the server where one is available.
  TextColumn get createdBy => text().nullable()();

  /// Who last edited the row.
  TextColumn get updatedBy => text().nullable()();

  /// The author's display name, denormalised so a record stays attributable
  /// without a join and without having downloaded the users table.
  TextColumn get createdByName => text().nullable()();
}

File: app/src/main/java/com/sankranthi/ledger/data/local/entity/SyncMeta.kt
package com.sankranthi.ledger.data.local.entity

/** Where a row stands with respect to the shared spreadsheet. */
enum class SyncStatus {
    /** Created or edited locally, not yet accepted by the gateway. */
    PENDING,

    /** The gateway has accepted this exact version. */
    SYNCED,

    /** Upload was rejected in a way retrying alone will not fix. */
    FAILED,

    /**
     * The gateway reported a newer server version than we based our edit on.
     * Needs reconciliation before it can be uploaded again.
     */
    CONFLICTED,
    ;

    val wire: String get() = name.lowercase()
}

/**
 * The sync bookkeeping every synchronised row carries, per §9 of the
 * requirements. Embedded rather than repeated so the rules stay in one place.
 *
 * Ownership matters here and is easy to get wrong:
 * - [version], [serverSeq] and [serverUpdatedAt] are **server-owned**. The client
 *   only ever copies back what the gateway returned.
 * - [createdAt] and [updatedAt] are client clocks and are for **display and
 *   ordering within this device only**. They are never used to resolve a
 *   conflict, because device clocks can be wrong or deliberately changed.
 */
data class SyncMeta(
    /** Server-assigned optimistic-concurrency counter. 0 until first accepted. */
    val version: Int = 0,

    /** Server-assigned monotonic sequence; also the download cursor. */
    val serverSeq: Long? = null,

    /** Authoritative timestamp from the gateway, ISO-8601. Null until synced. */
    val serverUpdatedAt: String? = null,

    val syncStatus: SyncStatus = SyncStatus.PENDING,

    /** Soft delete, so the removal can propagate to other devices. */
    val deleted: Boolean = false,

    val createdAt: Long = System.currentTimeMillis(),
    val updatedAt: Long = System.currentTimeMillis(),

    /** Email of the account that created / last edited the row. */
    val createdBy: String? = null,
    val updatedBy: String? = null,

    /** Display name of the creator, as the gateway reports it. For attribution UI. */
    val createdByName: String? = null,
) {
    val isSynced: Boolean get() = syncStatus == SyncStatus.SYNCED

    /** Marks the row as locally dirty again after an edit. */
    fun touched(by: String?, at: Long = System.currentTimeMillis()): SyncMeta =
        copy(syncStatus = SyncStatus.PENDING, updatedAt = at, updatedBy = by)
}
